library-build-transformer/src/main/kotlin/io/realm/buildtransformer/RealmBuildTransformer.kt

```diff
@@ -39,123 +52,124 @@ val logger: Logger = LoggerFactory.getLogger("realm-build-logger")
+    private fun transform() {
+        val outputProvider = JarOutputStream(
+            BufferedOutputStream(
+                FileOutputStream(
+                    output.get().asFile.absolutePath + ".tmp"
+                )
+            )
+        )
 
         val timer = Stopwatch()
         timer.start("Build Transform time")
-        if (isIncremental) {
-            runIncrementalTransform(inputs!!, outputProvider!!, transformClasses)
-        } else {
-            runFullTransform(inputs!!, outputProvider!!, transformClasses)
-        }
-        timer.stop()
-    }
 
-    private fun runFullTransform(inputs: MutableCollection<TransformInput>, outputProvider: TransformOutputProvider, transformClasses: Boolean) {
-        logger.debug("Run full transform")
-        val outputDir = outputProvider.getContentLocation("realmlibrarytransformer", outputTypes, scopes, Format.DIRECTORY)
-        val inputFiles: MutableSet<File> = mutableSetOf()
-        inputs.forEach {
-            it.directoryInputs.forEach {
-                // Non-incremental build: Include all files
-                val dirPath: String = it.file.absolutePath
-                it.file.walkTopDown()
-                        .filter { it.isFile }
-                        .filter { it.name.endsWith(".class") }
-                        .forEach { file ->
-                            file.packageHierarchyRootDir = dirPath
-                            file.shouldBeDeleted = transformClasses && specificFilesToStrip.find { file.absolutePath.endsWith(it) } != null
-                            inputFiles.add(file)
-                        }
+        val annotationDescriptor = createDescriptor(annotationQualifiedName.get())
+        val metadataCollector =
+            io.realm.buildtransformer.asm.visitors.AnnotationVisitor(annotationDescriptor)
+
+        forEachJarEntry { jarEntry, inputStream ->
+            if (jarEntry.name.endsWith(".class") ) {
+            inputStream.use {
+                val classReader = ClassReader(it)
+                classReader.accept(metadataCollector, 0)
+            }
             }
         }
-
-        transformClassFiles(outputDir, inputFiles, transformClasses)
-    }
-
-    private fun runIncrementalTransform(inputs: MutableCollection<TransformInput>, outputProvider: TransformOutputProvider, transformClasses: Boolean) {
-        logger.debug("Run incremental transform")
-        val outputDir = outputProvider.getContentLocation("realmlibrarytransformer", outputTypes, scopes, Format.DIRECTORY)
-        val inputFiles: MutableSet<File> = mutableSetOf()
-        inputs.forEach {
-            it.directoryInputs.forEach iterateDirs@{
-                if (!it.file.exists()) {
-                    return@iterateDirs // Directory was deleted
+        forEachJarEntry { jarEntry, inputStream ->
+            val bytes = inputStream.use { inputStream ->
+                if (jarEntry.name.endsWith(".class")) {
```

Comment:
What happens with ressource files in the JAR? We seem to skip them here, but I'm not sure how they then end up in the outputprovider?

Reply:
They are handled by the else branch 🎉 🤪